Migrate to Rust edition 2021

eduardosm opened the pull request:
<!-- Thank you for submitting a PR!

We have the contribution guide, please read it if you are new here!
<https://github.com/rust-lang/libc/blob/main/CONTRIBUTING.md>

Please fill out the below template.
-->

# Description

Migrates the libc crate to Rust edition 2021 with an automated script (after some initial manual adjustments).

Since ctest2 does not support `crate::` in paths, a hack has been implemented in `libc-test/build.rs`, which makes a copy of `src` replacing every `crate::` with `::`.

cc https://github.com/rust-lang/libc/issues/3898

<!-- Add a short description about what this change does -->

# Sources

<!-- All API changes must have links to headers and/or documentation,
preferably both -->

# Checklist

<!-- Please make sure the following has been done before submitting a PR,
or mark it as a draft if you are not sure. -->

- [x] Relevant tests in `libc-test/semver` have been updated
- [x] No placeholder or unstable values like `*LAST` or `*MAX` are
  included (see [#3131](https://github.com/rust-lang/libc/issues/3131))
- [ ] Tested locally (`cd libc-test && cargo test --target mytarget`);
  especially relevant for platforms that may not be checked in CI

bors commented:
:umbrella: The latest upstream changes (presumably #4128) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented on libc-test/build.rs:
```diff
@@ -145,11 +145,41 @@ fn main() {
+        std::fs::remove_dir_all("../target/src-hotfix").unwrap();
+    }
+
+    // ctest2 cannot parse `crate::` in paths, so replace them with `::`
```
Could you make this `// FIXME(ctest): ...` so it's greppable with the other ctest problems?

eduardosm replied:
Done

tgross35 commented:
Thanks! It is pretty unfortunate that the hotfix is needed (ctest is quite limiting...) but I think that is okay for now.

Would it be possible to add popular imports to the top of these files and then drop the `crate::` prefix? Not for everything of course, but it would be nice for the types that are used everywhere (all `c_*` types, `Ioctl`, `size_t`, `ssize_t`, anything else that is used somewhat commonly).

tgross35 commented:
Also`cargo fix` works with `--target` if you're up for just doing `rustc --print target-list | xargs -ITARGET cargo fix --edition --target TARGET --allow-dirty` (that's a bit too crude since some need build-std or downloaded targets and I don't expect every single target to actually work, but it should be possible to get everything that's in CI).

I'll also add a `.git-blame-ignore-revs` entry after this merges.

tgross35 commented:
> > Would it be possible to add popular imports to the top of these files and then drop the `crate::` prefix? Not for everything of course, but it would be nice for the types that are used everywhere (all `c_*` types, `Ioctl`, `size_t`, `ssize_t`, anything else that is used somewhat commonly).
> 
> Even with the help of automatic find-and-replace, that's going to be a pretty huge work considering the amount of modules where imports need to be added.

Cargo is aware of this right? Meaning if you put the import in the file, it will use that rather than adding the `crate::` prefix. If so, it should be pretty easy:

```sh
# `rg -l` lists files containing the pattern, the `sed` command
# will add the import to the top of the file
# The imports will be ugly but I'll just make rustfmt fix that
rg -l "::c_int" src | xargs -IFILE sed -i '1s;^;use crate::c_int\n\n;' FILE   
rg -l "::c_long" src | xargs -IFILE sed -i '1s;^;use crate::c_long\n\n;' FILE
# ... repeat for all relevant types

# fix commands
cargo fix ...

# run formatting
./ci/style.sh
```

Even if Cargo isn't aware it should be easy to sub out using the same `rg` command and replacing the sed pattern with `s/::c_int/c_int/g`.

Could you post the exact script you use to perform the automated step? Then have the commits in this PR be the following:

1. Add the testing hack
2. Any other setup, if applicable
3. Apply the automated commands from the script
4. Any hand cleanup, if applicable. (don't worry about updating the comments, so this can be smaller. Or I think you could do it in the script, duplicate the above `rg | xarg sed` commands and replace sed with with `perl -pi -e 's/(?!\w)::(\w+)/$1/g'`)

This will make it easy to verify the changes (I can rerun step 3 and hash the files to make sure nothing machine-specific and wonky happened - we caught a rustfmt bug this way when we reformatted part of rust-lang/rust), to rebase when the time comes (just drop commit 3 and rerun the script) and to apply to `libc-0.2` after this merges.

I am going to wait a couple days to merge this until it is possible to land a few changes that are nearly ready but would conflict with this, and also going to add rustfmt config so imports get grouped and sorted (meaning the output of the above script won't be horrible). So feel free to hold off on rebasing unless you're experimenting with the above, I'll let you know when I'm ready for this 👍

eduardosm commented:
I'm seeing some weirdness:

If I add `use crate::c_long;` to `src/unix/linux_like/linux/mod.rs`, it fails to compile with
```
error[E0432]: unresolved import `crate::c_long`
 --> src/unix/linux_like/linux/mod.rs:3:5
  |
3 | use crate::c_long;
  |     ^^^^^^^^^^^^^ no `c_long` in the root
```

However, if I just change a use of `::c_long` with `crate::c_long` anywhere in the same file, it compiles fine.

What's going on there? Is that a rustc bug?

tgross35 commented:
Oh - I guess that's actually a new feature in 2018 https://doc.rust-lang.org/edition-guide/rust-2018/path-changes.html#the-crate-keyword-refers-to-the-current-crate. After the changes here, are you able to actually set the edition to 2018? Or are there other failures.

eduardosm commented:
But `use crate::` usually works on edition 2015 too, since it does not conflict with any previous syntax. If fact, `use crate::c_int;` works fine. It is with `c_long` where I'm getting errors.

tgross35 commented:
Here's a repro - I guess you can't reexport something through a module and then also `use` it from the crate https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e124499dd4f73870bdddd5e64eadfaf2, `c_int` is different because it's at a higher level.

I'm okay with just leaving the `crate::` prefix on `c_long` and any others that give you problems, I'll figure out how to clean them up later (unless you feel like hacking at it). As long as we get most of the `crate::` prefixes in the automated step so the diff isn't bigger than it needs to be.

eduardosm commented:
Is there a reason not to use types from `core::ffi`?

tgross35 commented:
The `core::ffi` types became stable in 1.64, which is beyond our current MSRV, and we can't use `std` for `os::raw`. This would be a good thing to change at the next MSRV bump though.

eduardosm commented:
`cargo fix` does not seem to be that smart. It will change `::<type>` to `crate::<type>` even if there is a `use crate::<type>`.

I'll try to hack some find-and-replace solution.

tgross35 commented:
Well that's interesting, I thought it would know better. Here's one solution (`(?<!\w)` ensures the `::` is not preceded by another word character, i.e. should match ` ::foo` but not `crate::foo`)

```
find src -name '*.rs' | xargs -IFILE perl -pi -077 -e \
    's/(?<!\w)::(c_[us]?char|c_double|c_float|c_u?int|c_u?longlong|c_u?short|ptrdiff_t|s?size_t|c_void|Ioctl|off_t|time_t)/$1/g' FILE
```

bors commented:
:umbrella: The latest upstream changes (presumably #4136) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #3978) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

eduardosm commented:
Now there are 3 commits instead of 4. The changes of the old third commit are now also handled by the script.

<details>

<summary>New script</summary>

```sh
#!/usr/bin/env bash

git restore Cargo.toml src

# Handle leading `::` in common types

implicit_c_long_ulong_char=(
  src/fuchsia/mod.rs
  src/solid/mod.rs
  src/unix/aix/mod.rs
  src/unix/bsd/apple/mod.rs
  src/unix/bsd/freebsdlike/freebsd/mod.rs
  src/unix/bsd/freebsdlike/mod.rs
  src/unix/bsd/netbsdlike/netbsd/mod.rs
  src/unix/bsd/netbsdlike/openbsd/mod.rs
  src/unix/bsd/netbsdlike/mod.rs
  src/unix/bsd/mod.rs
  src/unix/haiku/mod.rs
  src/unix/hurd/mod.rs
  src/unix/linux_like/android/b32/mod.rs
  src/unix/linux_like/android/b64/mod.rs
  src/unix/linux_like/android/mod.rs
  src/unix/linux_like/linux/gnu/b32/mod.rs
  src/unix/linux_like/linux/gnu/b64/aarch64/mod.rs
  src/unix/linux_like/linux/gnu/b64/x86_64/mod.rs
  src/unix/linux_like/linux/gnu/b64/mod.rs
  src/unix/linux_like/linux/gnu/mod.rs
  src/unix/linux_like/linux/musl/mod.rs
  src/unix/linux_like/linux/uclibc/mips/mod.rs
  src/unix/linux_like/linux/uclibc/mod.rs
  src/unix/linux_like/linux/mod.rs
  src/unix/linux_like/mod.rs
  src/unix/newlib/mod.rs
  src/unix/nto/mod.rs
  src/unix/mod.rs
  src/vxworks/mod.rs
)

for file in "${implicit_c_long_ulong_char[@]}"; do
  perl -pi -077 -e 's/(?<!\w)::(c_long|c_ulong|c_char)\b/$1/g' "$file"
done

implicit_off_t=(
  src/unix/linux_like/linux/gnu/mod.rs
  src/unix/linux_like/linux/mod.rs
  src/unix/linux_like/mod.rs
  src/unix/mod.rs
)

for file in "${implicit_off_t[@]}"; do
  perl -pi -077 -e 's/(?<!\w)::(off_t)\b/$1/g' "$file"
done

implicit_off64_t=(
  src/unix/linux_like/android/mod.rs
  src/unix/linux_like/mod.rs
)

for file in "${implicit_off64_t[@]}"; do
  perl -pi -077 -e 's/(?<!\w)::(off64_t)\b/$1/g' "$file"
done

implicit_Ioctl=(
  src/unix/linux_like/linux/mod.rs
)

for file in "${implicit_Ioctl[@]}"; do
  perl -pi -077 -e 's/(?<!\w)::(Ioctl)\b/$1/g' "$file"
done

perl -pi -077 -e 's/(?<!\w)::(__u8|__u16|__u32|__s16|__s32)\b/$1/g' src/unix/linux_like/linux/mod.rs

perl -pi -077 -e 's/(?<!\w)::(SIG[A-Z]+)\b/$1/g' src/unix/solarish/mod.rs

# These are used only in `cfg`ed items or comments, `use ::<type>;` would cause warnings on some targets
perl -pi -077 -e 's/(?<!\w)::c_ulong\b/crate::c_ulong/g' src/unix/newlib/generic.rs
perl -pi -077 -e 's/(?<!\w)::c_ulong\b/crate::c_ulong/g' src/unix/linux_like/linux/arch/generic/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_ulonglong\b/crate::c_ulonglong/g' src/unix/linux_like/linux/gnu/b64/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_ulonglong\b/crate::c_ulonglong/g' src/unix/linux_like/linux/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_ulonglong\b/crate::c_ulonglong/g' src/unix/linux_like/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_short\b/crate::c_short/g' src/unix/linux_like/linux/gnu/b32/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_short\b/crate::c_short/g' src/unix/newlib/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_float\b/c_float/g' src/unix/linux_like/linux/gnu/b64/s390x.rs
perl -pi -077 -e 's/(?<!\w)::c_float\b/c_float/g' src/unix/linux_like/linux/musl/b64/s390x.rs

function handle_type() {
  type="$1"
  uses_type="$(rg -l "[^\w]::$type\b" src)"
  has_type_def="$(rg -l "(type|struct) $type " src)"
  has_type_import="$(rg -l "use ([0-9a-zA-Z_]+::)*$type;" src)"

  add_import=($(comm -3 <(echo "$uses_type" | sort) <(printf "%s\n%s" "$has_type_def" "$has_type_import" | sort) | cut -f1))
  for file in "${add_import[@]}"; do
    # Split the file into two parts: module-level attributes and rest
    # Imports will be added after module-level attributes

    awk '/^#!|^\/\/!/ {found=NR} {lines[NR]=$0} END {for (i=1; i<=found; i++) print lines[i]}' \
      "$file" > "$file.attrs"
    awk '/^#!|^\/\/!/ {found=NR} END {if (found) {for (i=found+1; i<=NR; i++) print lines[i]} else {for (i=1; i<=NR; i++) print lines[i]}} {lines[NR]=$0}' \
      "$file" > "$file.rest"

    cat "$file.attrs" > "$file"
    echo >> "$file"
    echo 'use crate::'"$type"';' >> "$file"
    echo >> "$file"
    cat "$file.rest" >> "$file"

    rm "$file.attrs" "$file.rest"
  done

  for file in $uses_type; do
    perl -pi -077 -e 's/(?<!\w)::('"$type"')\b/$1/g' "$file"
  done
}

handle_type c_void
handle_type c_int
handle_type c_uint
handle_type c_char
handle_type c_schar
handle_type c_uchar
handle_type c_short
handle_type c_ushort
handle_type c_long
handle_type c_ulong
handle_type c_longlong
handle_type c_ulonglong
handle_type c_float
handle_type c_double
handle_type size_t
handle_type ssize_t
handle_type intptr_t
handle_type ptrdiff_t
handle_type off_t
handle_type off64_t
#handle_type time_t # Produces deprecation warnings
handle_type Ioctl
handle_type cmsghdr # To avoid "line formatted, but exceeded maximum width"

# Commented targets failed to `cargo check -Zbuild-std=core`
targets=(
  aarch64-apple-darwin
  aarch64-apple-ios
  aarch64-apple-ios-macabi
  aarch64-apple-ios-sim
  aarch64-apple-tvos
  aarch64-apple-tvos-sim
  aarch64-apple-visionos
  aarch64-apple-visionos-sim
  aarch64-apple-watchos
  aarch64-apple-watchos-sim
  aarch64-kmc-solid_asp3
  aarch64-linux-android
  aarch64-nintendo-switch-freestanding
  aarch64-pc-windows-gnullvm
  aarch64-pc-windows-msvc
  aarch64-unknown-freebsd
  aarch64-unknown-fuchsia
  aarch64-unknown-hermit
  aarch64-unknown-illumos
  aarch64-unknown-linux-gnu
  aarch64-unknown-linux-gnu_ilp32
  aarch64-unknown-linux-musl
  aarch64-unknown-linux-ohos
  aarch64-unknown-netbsd
  aarch64-unknown-none
  aarch64-unknown-none-softfloat
  aarch64-unknown-nto-qnx700
  aarch64-unknown-nto-qnx710
  aarch64-unknown-openbsd
  aarch64-unknown-redox
  #aarch64-unknown-teeos
  aarch64-unknown-trusty
  aarch64-unknown-uefi
  aarch64-uwp-windows-msvc
  aarch64-wrs-vxworks
  aarch64_be-unknown-linux-gnu
  aarch64_be-unknown-linux-gnu_ilp32
  aarch64_be-unknown-netbsd
  arm-linux-androideabi
  arm-unknown-linux-gnueabi
  arm-unknown-linux-gnueabihf
  arm-unknown-linux-musleabi
  arm-unknown-linux-musleabihf
  arm64_32-apple-watchos
  arm64e-apple-darwin
  arm64e-apple-ios
  arm64e-apple-tvos
  arm64ec-pc-windows-msvc
  armeb-unknown-linux-gnueabi
  armebv7r-none-eabi
  armebv7r-none-eabihf
  armv4t-none-eabi
  armv4t-unknown-linux-gnueabi
  armv5te-none-eabi
  armv5te-unknown-linux-gnueabi
  armv5te-unknown-linux-musleabi
  armv5te-unknown-linux-uclibceabi
  armv6-unknown-freebsd
  armv6-unknown-netbsd-eabihf
  armv6k-nintendo-3ds
  armv7-linux-androideabi
  armv7-rtems-eabihf
  armv7-sony-vita-newlibeabihf
  armv7-unknown-freebsd
  armv7-unknown-linux-gnueabi
  armv7-unknown-linux-gnueabihf
  armv7-unknown-linux-musleabi
  armv7-unknown-linux-musleabihf
  armv7-unknown-linux-ohos
  armv7-unknown-linux-uclibceabi
  armv7-unknown-linux-uclibceabihf
  armv7-unknown-netbsd-eabihf
  armv7-unknown-trusty
  armv7-wrs-vxworks-eabihf
  armv7a-kmc-solid_asp3-eabi
  armv7a-kmc-solid_asp3-eabihf
  armv7a-none-eabi
  armv7a-none-eabihf
  armv7k-apple-watchos
  armv7r-none-eabi
  armv7r-none-eabihf
  armv7s-apple-ios
  armv8r-none-eabihf
  #avr-unknown-gnu-atmega328
  bpfeb-unknown-none
  bpfel-unknown-none
  csky-unknown-linux-gnuabiv2
  csky-unknown-linux-gnuabiv2hf
  hexagon-unknown-linux-musl
  #hexagon-unknown-none-elf
  i386-apple-ios
  #i586-pc-nto-qnx700
  i586-pc-windows-msvc
  i586-unknown-linux-gnu
  i586-unknown-linux-musl
  i586-unknown-netbsd
  i686-apple-darwin
  i686-linux-android
  i686-pc-windows-gnu
  i686-pc-windows-gnullvm
  i686-pc-windows-msvc
  i686-unknown-freebsd
  i686-unknown-haiku
  i686-unknown-hurd-gnu
  i686-unknown-linux-gnu
  i686-unknown-linux-musl
  i686-unknown-netbsd
  i686-unknown-openbsd
  i686-unknown-redox
  i686-unknown-uefi
  i686-uwp-windows-gnu
  i686-uwp-windows-msvc
  i686-win7-windows-msvc
  i686-wrs-vxworks
  loongarch64-unknown-linux-gnu
  loongarch64-unknown-linux-musl
  loongarch64-unknown-linux-ohos
  loongarch64-unknown-none
  loongarch64-unknown-none-softfloat
  #m68k-unknown-linux-gnu
  mips-unknown-linux-gnu
  mips-unknown-linux-musl
  mips-unknown-linux-uclibc
  mips64-openwrt-linux-musl
  mips64-unknown-linux-gnuabi64
  mips64-unknown-linux-muslabi64
  mips64el-unknown-linux-gnuabi64
  mips64el-unknown-linux-muslabi64
  mipsel-sony-psp
  mipsel-sony-psx
  mipsel-unknown-linux-gnu
  mipsel-unknown-linux-musl
  mipsel-unknown-linux-uclibc
  mipsel-unknown-netbsd
  mipsel-unknown-none
  mipsisa32r6-unknown-linux-gnu
  mipsisa32r6el-unknown-linux-gnu
  mipsisa64r6-unknown-linux-gnuabi64
  mipsisa64r6el-unknown-linux-gnuabi64
  msp430-none-elf
  nvptx64-nvidia-cuda
  powerpc-unknown-freebsd
  powerpc-unknown-linux-gnu
  powerpc-unknown-linux-gnuspe
  powerpc-unknown-linux-musl
  powerpc-unknown-linux-muslspe
  powerpc-unknown-netbsd
  powerpc-unknown-openbsd
  powerpc-wrs-vxworks
  powerpc-wrs-vxworks-spe
  powerpc64-ibm-aix
  powerpc64-unknown-freebsd
  powerpc64-unknown-linux-gnu
  powerpc64-unknown-linux-musl
  powerpc64-unknown-openbsd
  powerpc64-wrs-vxworks
  powerpc64le-unknown-freebsd
  powerpc64le-unknown-linux-gnu
  powerpc64le-unknown-linux-musl
  riscv32-wrs-vxworks
  riscv32e-unknown-none-elf
  riscv32em-unknown-none-elf
  riscv32emc-unknown-none-elf
  riscv32gc-unknown-linux-gnu
  riscv32gc-unknown-linux-musl
  riscv32i-unknown-none-elf
  riscv32im-risc0-zkvm-elf
  riscv32im-unknown-none-elf
  riscv32ima-unknown-none-elf
  riscv32imac-esp-espidf
  riscv32imac-unknown-none-elf
  riscv32imac-unknown-nuttx-elf
  riscv32imac-unknown-xous-elf
  riscv32imafc-esp-espidf
  riscv32imafc-unknown-none-elf
  riscv32imafc-unknown-nuttx-elf
  riscv32imc-esp-espidf
  riscv32imc-unknown-none-elf
  riscv32imc-unknown-nuttx-elf
  riscv64-linux-android
  riscv64-wrs-vxworks
  riscv64gc-unknown-freebsd
  riscv64gc-unknown-fuchsia
  riscv64gc-unknown-hermit
  riscv64gc-unknown-linux-gnu
  riscv64gc-unknown-linux-musl
  #riscv64gc-unknown-netbsd
  riscv64gc-unknown-none-elf
  riscv64gc-unknown-nuttx-elf
  riscv64gc-unknown-openbsd
  riscv64imac-unknown-none-elf
  riscv64imac-unknown-nuttx-elf
  s390x-unknown-linux-gnu
  s390x-unknown-linux-musl
  sparc-unknown-linux-gnu
  sparc-unknown-none-elf
  sparc64-unknown-linux-gnu
  sparc64-unknown-netbsd
  sparc64-unknown-openbsd
  sparcv9-sun-solaris
  thumbv4t-none-eabi
  thumbv5te-none-eabi
  thumbv6m-none-eabi
  thumbv6m-nuttx-eabi
  thumbv7a-pc-windows-msvc
  thumbv7a-uwp-windows-msvc
  thumbv7em-none-eabi
  thumbv7em-none-eabihf
  thumbv7em-nuttx-eabi
  thumbv7em-nuttx-eabihf
  thumbv7m-none-eabi
  thumbv7m-nuttx-eabi
  thumbv7neon-linux-androideabi
  thumbv7neon-unknown-linux-gnueabihf
  thumbv7neon-unknown-linux-musleabihf
  thumbv8m.base-none-eabi
  thumbv8m.base-nuttx-eabi
  thumbv8m.main-none-eabi
  thumbv8m.main-none-eabihf
  thumbv8m.main-nuttx-eabi
  thumbv8m.main-nuttx-eabihf
  wasm32-unknown-emscripten
  wasm32-unknown-unknown
  wasm32-wasip1
  wasm32-wasip1-threads
  wasm32-wasip2
  wasm32v1-none
  wasm64-unknown-unknown
  x86_64-apple-darwin
  x86_64-apple-ios
  x86_64-apple-ios-macabi
  x86_64-apple-tvos
  x86_64-apple-watchos-sim
  x86_64-fortanix-unknown-sgx
  x86_64-linux-android
  x86_64-pc-nto-qnx710
  x86_64-pc-solaris
  x86_64-pc-windows-gnu
  x86_64-pc-windows-gnullvm
  x86_64-pc-windows-msvc
  x86_64-unikraft-linux-musl
  x86_64-unknown-dragonfly
  x86_64-unknown-freebsd
  x86_64-unknown-fuchsia
  x86_64-unknown-haiku
  x86_64-unknown-hermit
  x86_64-unknown-hurd-gnu
  x86_64-unknown-illumos
  x86_64-unknown-l4re-uclibc
  x86_64-unknown-linux-gnu
  x86_64-unknown-linux-gnux32
  x86_64-unknown-linux-musl
  #x86_64-unknown-linux-none
  x86_64-unknown-linux-ohos
  x86_64-unknown-netbsd
  x86_64-unknown-none
  x86_64-unknown-openbsd
  x86_64-unknown-redox
  x86_64-unknown-trusty
  x86_64-unknown-uefi
  x86_64-uwp-windows-gnu
  x86_64-uwp-windows-msvc
  x86_64-win7-windows-msvc
  x86_64-wrs-vxworks
  x86_64h-apple-darwin
  #xtensa-esp32-espidf
  #xtensa-esp32-none-elf
  #xtensa-esp32s2-espidf
  #xtensa-esp32s2-none-elf
  #xtensa-esp32s3-espidf
  #xtensa-esp32s3-none-elf
)

# Migrate to 2018

for target in "${targets[@]}"; do
  echo "Migrating 2015 -> 2018 target $target"
  cargo fix --target "$target" -Zbuild-std=core --features extra_traits --allow-dirty --edition --broken-code 2> /dev/null
done

# Handle freebsd versions
freebsd_versions=(freebsd13 freebsd14 freebsd15)
for ver in "${freebsd_versions[@]}"; do
  echo "Migrating 2015 -> 2018 $ver"
  RUSTFLAGS="--cfg $ver" cargo fix --target i686-unknown-freebsd -Zbuild-std=core --features extra_traits --allow-dirty --edition --broken-code 2> /dev/null
  RUSTFLAGS="--cfg $ver" cargo fix --target x86_64-unknown-freebsd -Zbuild-std=core --features extra_traits --allow-dirty --edition --broken-code 2> /dev/null
done

sed -i 's/edition = "2015"/edition = "2018"/' Cargo.toml

# Migrate to 2021

for target in "${targets[@]}"; do
  echo "Migrating 2018 -> 2021 target $target"
  cargo fix --target "$target" -Zbuild-std=core --features extra_traits --allow-dirty --edition --broken-code 2> /dev/null
done

sed -i 's/edition = "2018"/edition = "2021"/' Cargo.toml

# Check
for target in "${targets[@]}"; do
  echo "Checking target $target"
  cargo check --target "$target" -Zbuild-std=core --features extra_traits || {
    echo "Failed for target $target"
    exit 1
  }
done

for ver in "${freebsd_versions[@]}"; do
  echo "Checking $ver"
  RUSTFLAGS="--cfg $ver" cargo check --target i686-unknown-freebsd -Zbuild-std=core --features extra_traits || {
    echo "Failed for i686 $freebsd_versions"
    exit 1
  }
  RUSTFLAGS="--cfg $ver" cargo check --target x86_64-unknown-freebsd -Zbuild-std=core --features extra_traits || {
    echo "Failed for x86_64 $freebsd_versions"
    exit 1
  }
done

# Fix formatting
ci/style.sh
```

</details>

tgross35 commented on src/unix/bsd/freebsdlike/freebsd/mod.rs:
```diff
@@ -57,8 +61,8 @@ pub enum devstat_support_flags {
 }
-impl ::Copy for devstat_support_flags {}
-impl ::Clone for devstat_support_flags {
+impl crate::Copy for devstat_support_flags {}
+impl crate::Clone for devstat_support_flags {
```
These are in the prelude so the script could strip the `crate::` prefix

tgross35 commented on src/unix/bsd/apple/b64/x86_64/mod.rs:
```diff
@@ -91,82 +93,82 @@ s! {
         >,
-        pub malloc: ::Option<
-            unsafe extern "C" fn(zone: *mut malloc_zone_t, size: ::size_t) -> *mut ::c_void,
+        pub malloc: crate::Option<
```
The script could replace `crate::Option` with `Option`

tgross35 commented:
Results of me running locally:

```console
$ git checkout 498165de32e5c5661b58f108ff32a7ef75454534 # last commit in this PR before automated upgrades
$ git ls-tree --full-tree -r HEAD . | git hash-object  --stdin # hash of initial state
e210fdd4cb497b8cb13c2b962ea0de4b6f5ba2a5
$ ../a.sh  # your script
$ git commit -am "apply automated changes" # dummy commit for ls-tree
$ git ls-tree --full-tree -r HEAD . | git hash-object --stdin # hash the result of these changes
72db53fcc2e536e8725f4178d035bace1f853a3d
$  git diff dfb720856c05943c34085d541a235a466dc9d751 # compare to the result of this PR
diff --git a/Cargo.toml b/Cargo.toml
index a631b4a5d..9deb2ac56 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -4,7 +4,7 @@ version = "1.0.0-alpha.1"
 authors = ["The Rust Project Developers"]
 license = "MIT OR Apache-2.0"
 readme = "README.md"
-edition = "2021"
+edition = "2015"
 repository = "https://github.com/rust-lang/libc"
 homepage = "https://github.com/rust-lang/libc"
 documentation = "https://docs.rs/libc/"
diff --git a/src/unix/linux_like/mod.rs b/src/unix/linux_like/mod.rs
index 823a4ec02..ac17f88f2 100644
--- a/src/unix/linux_like/mod.rs
+++ b/src/unix/linux_like/mod.rs
@@ -1658,7 +1658,7 @@ safe_f! {
     pub {const} fn KERNEL_VERSION(a: u32, b: u32, c: u32) -> u32 {
         ((a << 16) + (b << 8))
             + match c {
-                0..=255 => c,
+                0...255 => c,
                 _ => 255,
             }
     }
$ git checkout dfb720856c05943c34085d541a235a466dc9d751 # commit in this PR after automated changes
$ git ls-tree --full-tree -r HEAD . | git hash-object --stdin
e0aaa981dce61d44bcf37469c58e9aae5ba0011c
```

I'm not sure why the above would be different?

_Note to self so I don't accidentally invalidate 30GB of cache: I'm on `rustc 1.84.0-nightly (b19329a37 2024-11-21)`_

tgross35 commented:
Anyway I'm not going to worry about the above, it looks like your script is doing the right thing and my machine isn't for some reason. Everything else lines up so that is the important thing.

I left two comments suggesting some trivial updates. Rebase and rerun whenever you're ready, then let's get this huge thing merged!

tgross35 commented:
> I'm not sure why the above would be different?

Oh, the edition string update was done with `sed` which isn't portable. If those sed lines were replaced with awk or perl it would probably be fine.

eduardosm commented:
I didn't know that `sed -i` was not portable. I replaced it with perl. Rebased and removed `::` prefixes for `Option`, `Copy` and `Clone`.

<details>

<summary>Script</summary>

```sh
#!/usr/bin/env bash

git restore Cargo.toml src

# Handle leading `::` in common types

for file in $(find src -name "*.rs"); do
  perl -pi -077 -e 's/(?<!\w)::(Option|Copy|Clone)\b/$1/g' "$file"
done

implicit_c_long_ulong_char=(
  src/fuchsia/mod.rs
  src/solid/mod.rs
  src/unix/aix/mod.rs
  src/unix/bsd/apple/mod.rs
  src/unix/bsd/freebsdlike/freebsd/mod.rs
  src/unix/bsd/freebsdlike/mod.rs
  src/unix/bsd/netbsdlike/netbsd/mod.rs
  src/unix/bsd/netbsdlike/openbsd/mod.rs
  src/unix/bsd/netbsdlike/mod.rs
  src/unix/bsd/mod.rs
  src/unix/haiku/mod.rs
  src/unix/hurd/mod.rs
  src/unix/linux_like/android/b32/mod.rs
  src/unix/linux_like/android/b64/mod.rs
  src/unix/linux_like/android/mod.rs
  src/unix/linux_like/linux/gnu/b32/mod.rs
  src/unix/linux_like/linux/gnu/b64/aarch64/mod.rs
  src/unix/linux_like/linux/gnu/b64/x86_64/mod.rs
  src/unix/linux_like/linux/gnu/b64/mod.rs
  src/unix/linux_like/linux/gnu/mod.rs
  src/unix/linux_like/linux/musl/mod.rs
  src/unix/linux_like/linux/uclibc/mips/mod.rs
  src/unix/linux_like/linux/uclibc/mod.rs
  src/unix/linux_like/linux/mod.rs
  src/unix/linux_like/mod.rs
  src/unix/newlib/mod.rs
  src/unix/nto/mod.rs
  src/unix/mod.rs
  src/vxworks/mod.rs
)

for file in "${implicit_c_long_ulong_char[@]}"; do
  perl -pi -077 -e 's/(?<!\w)::(c_long|c_ulong|c_char)\b/$1/g' "$file"
done

implicit_off_t=(
  src/unix/linux_like/linux/gnu/mod.rs
  src/unix/linux_like/linux/mod.rs
  src/unix/linux_like/mod.rs
  src/unix/mod.rs
)

for file in "${implicit_off_t[@]}"; do
  perl -pi -077 -e 's/(?<!\w)::(off_t)\b/$1/g' "$file"
done

implicit_off64_t=(
  src/unix/linux_like/android/mod.rs
  src/unix/linux_like/mod.rs
)

for file in "${implicit_off64_t[@]}"; do
  perl -pi -077 -e 's/(?<!\w)::(off64_t)\b/$1/g' "$file"
done

implicit_Ioctl=(
  src/unix/linux_like/linux/mod.rs
)

for file in "${implicit_Ioctl[@]}"; do
  perl -pi -077 -e 's/(?<!\w)::(Ioctl)\b/$1/g' "$file"
done

perl -pi -077 -e 's/(?<!\w)::(__u8|__u16|__u32|__s16|__s32)\b/$1/g' src/unix/linux_like/linux/mod.rs

perl -pi -077 -e 's/(?<!\w)::(SIG[A-Z]+)\b/$1/g' src/unix/solarish/mod.rs

# These are used only in `cfg`ed items or comments, `use ::<type>;` would cause warnings on some targets
perl -pi -077 -e 's/(?<!\w)::c_ulong\b/crate::c_ulong/g' src/unix/newlib/generic.rs
perl -pi -077 -e 's/(?<!\w)::c_ulong\b/crate::c_ulong/g' src/unix/linux_like/linux/arch/generic/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_ulonglong\b/crate::c_ulonglong/g' src/unix/linux_like/linux/gnu/b64/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_ulonglong\b/crate::c_ulonglong/g' src/unix/linux_like/linux/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_ulonglong\b/crate::c_ulonglong/g' src/unix/linux_like/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_short\b/crate::c_short/g' src/unix/linux_like/linux/gnu/b32/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_short\b/crate::c_short/g' src/unix/newlib/mod.rs
perl -pi -077 -e 's/(?<!\w)::c_float\b/c_float/g' src/unix/linux_like/linux/gnu/b64/s390x.rs
perl -pi -077 -e 's/(?<!\w)::c_float\b/c_float/g' src/unix/linux_like/linux/musl/b64/s390x.rs

function handle_type() {
  type="$1"
  uses_type="$(rg -l "[^\w]::$type\b" src)"
  has_type_def="$(rg -l "(type|struct) $type " src)"
  has_type_import="$(rg -l "use ([0-9a-zA-Z_]+::)*$type;" src)"

  add_import=($(comm -3 <(echo "$uses_type" | sort) <(printf "%s\n%s" "$has_type_def" "$has_type_import" | sort) | cut -f1))
  for file in "${add_import[@]}"; do
    # Split the file into two parts: module-level attributes and rest
    # Imports will be added after module-level attributes

    awk '/^#!|^\/\/!/ {found=NR} {lines[NR]=$0} END {for (i=1; i<=found; i++) print lines[i]}' \
      "$file" > "$file.attrs"
    awk '/^#!|^\/\/!/ {found=NR} END {if (found) {for (i=found+1; i<=NR; i++) print lines[i]} else {for (i=1; i<=NR; i++) print lines[i]}} {lines[NR]=$0}' \
      "$file" > "$file.rest"

    cat "$file.attrs" > "$file"
    echo >> "$file"
    echo 'use crate::'"$type"';' >> "$file"
    echo >> "$file"
    cat "$file.rest" >> "$file"

    rm "$file.attrs" "$file.rest"
  done

  for file in $uses_type; do
    perl -pi -077 -e 's/(?<!\w)::('"$type"')\b/$1/g' "$file"
  done
}

handle_type c_void
handle_type c_int
handle_type c_uint
handle_type c_char
handle_type c_schar
handle_type c_uchar
handle_type c_short
handle_type c_ushort
handle_type c_long
handle_type c_ulong
handle_type c_longlong
handle_type c_ulonglong
handle_type c_float
handle_type c_double
handle_type size_t
handle_type ssize_t
handle_type intptr_t
handle_type ptrdiff_t
handle_type off_t
handle_type off64_t
#handle_type time_t # Produces deprecation warnings
handle_type Ioctl
handle_type cmsghdr # To avoid "line formatted, but exceeded maximum width"

# Commented targets failed to `cargo check -Zbuild-std=core`
targets=(
  aarch64-apple-darwin
  aarch64-apple-ios
  aarch64-apple-ios-macabi
  aarch64-apple-ios-sim
  aarch64-apple-tvos
  aarch64-apple-tvos-sim
  aarch64-apple-visionos
  aarch64-apple-visionos-sim
  aarch64-apple-watchos
  aarch64-apple-watchos-sim
  aarch64-kmc-solid_asp3
  aarch64-linux-android
  aarch64-nintendo-switch-freestanding
  aarch64-pc-windows-gnullvm
  aarch64-pc-windows-msvc
  aarch64-unknown-freebsd
  aarch64-unknown-fuchsia
  aarch64-unknown-hermit
  aarch64-unknown-illumos
  aarch64-unknown-linux-gnu
  aarch64-unknown-linux-gnu_ilp32
  aarch64-unknown-linux-musl
  aarch64-unknown-linux-ohos
  aarch64-unknown-netbsd
  aarch64-unknown-none
  aarch64-unknown-none-softfloat
  aarch64-unknown-nto-qnx700
  aarch64-unknown-nto-qnx710
  aarch64-unknown-openbsd
  aarch64-unknown-redox
  #aarch64-unknown-teeos
  aarch64-unknown-trusty
  aarch64-unknown-uefi
  aarch64-uwp-windows-msvc
  aarch64-wrs-vxworks
  aarch64_be-unknown-linux-gnu
  aarch64_be-unknown-linux-gnu_ilp32
  aarch64_be-unknown-netbsd
  arm-linux-androideabi
  arm-unknown-linux-gnueabi
  arm-unknown-linux-gnueabihf
  arm-unknown-linux-musleabi
  arm-unknown-linux-musleabihf
  arm64_32-apple-watchos
  arm64e-apple-darwin
  arm64e-apple-ios
  arm64e-apple-tvos
  arm64ec-pc-windows-msvc
  armeb-unknown-linux-gnueabi
  armebv7r-none-eabi
  armebv7r-none-eabihf
  armv4t-none-eabi
  armv4t-unknown-linux-gnueabi
  armv5te-none-eabi
  armv5te-unknown-linux-gnueabi
  armv5te-unknown-linux-musleabi
  armv5te-unknown-linux-uclibceabi
  armv6-unknown-freebsd
  armv6-unknown-netbsd-eabihf
  armv6k-nintendo-3ds
  armv7-linux-androideabi
  armv7-rtems-eabihf
  armv7-sony-vita-newlibeabihf
  armv7-unknown-freebsd
  armv7-unknown-linux-gnueabi
  armv7-unknown-linux-gnueabihf
  armv7-unknown-linux-musleabi
  armv7-unknown-linux-musleabihf
  armv7-unknown-linux-ohos
  armv7-unknown-linux-uclibceabi
  armv7-unknown-linux-uclibceabihf
  armv7-unknown-netbsd-eabihf
  armv7-unknown-trusty
  armv7-wrs-vxworks-eabihf
  armv7a-kmc-solid_asp3-eabi
  armv7a-kmc-solid_asp3-eabihf
  armv7a-none-eabi
  armv7a-none-eabihf
  armv7k-apple-watchos
  armv7r-none-eabi
  armv7r-none-eabihf
  armv7s-apple-ios
  armv8r-none-eabihf
  #avr-unknown-gnu-atmega328
  bpfeb-unknown-none
  bpfel-unknown-none
  csky-unknown-linux-gnuabiv2
  csky-unknown-linux-gnuabiv2hf
  hexagon-unknown-linux-musl
  #hexagon-unknown-none-elf
  i386-apple-ios
  #i586-pc-nto-qnx700
  i586-pc-windows-msvc
  i586-unknown-linux-gnu
  i586-unknown-linux-musl
  i586-unknown-netbsd
  i686-apple-darwin
  i686-linux-android
  i686-pc-windows-gnu
  i686-pc-windows-gnullvm
  i686-pc-windows-msvc
  i686-unknown-freebsd
  i686-unknown-haiku
  i686-unknown-hurd-gnu
  i686-unknown-linux-gnu
  i686-unknown-linux-musl
  i686-unknown-netbsd
  i686-unknown-openbsd
  i686-unknown-redox
  i686-unknown-uefi
  i686-uwp-windows-gnu
  i686-uwp-windows-msvc
  i686-win7-windows-msvc
  i686-wrs-vxworks
  loongarch64-unknown-linux-gnu
  loongarch64-unknown-linux-musl
  loongarch64-unknown-linux-ohos
  loongarch64-unknown-none
  loongarch64-unknown-none-softfloat
  #m68k-unknown-linux-gnu
  mips-unknown-linux-gnu
  mips-unknown-linux-musl
  mips-unknown-linux-uclibc
  mips64-openwrt-linux-musl
  mips64-unknown-linux-gnuabi64
  mips64-unknown-linux-muslabi64
  mips64el-unknown-linux-gnuabi64
  mips64el-unknown-linux-muslabi64
  mipsel-sony-psp
  mipsel-sony-psx
  mipsel-unknown-linux-gnu
  mipsel-unknown-linux-musl
  mipsel-unknown-linux-uclibc
  mipsel-unknown-netbsd
  mipsel-unknown-none
  mipsisa32r6-unknown-linux-gnu
  mipsisa32r6el-unknown-linux-gnu
  mipsisa64r6-unknown-linux-gnuabi64
  mipsisa64r6el-unknown-linux-gnuabi64
  msp430-none-elf
  nvptx64-nvidia-cuda
  powerpc-unknown-freebsd
  powerpc-unknown-linux-gnu
  powerpc-unknown-linux-gnuspe
  powerpc-unknown-linux-musl
  powerpc-unknown-linux-muslspe
  powerpc-unknown-netbsd
  powerpc-unknown-openbsd
  powerpc-wrs-vxworks
  powerpc-wrs-vxworks-spe
  powerpc64-ibm-aix
  powerpc64-unknown-freebsd
  powerpc64-unknown-linux-gnu
  powerpc64-unknown-linux-musl
  powerpc64-unknown-openbsd
  powerpc64-wrs-vxworks
  powerpc64le-unknown-freebsd
  powerpc64le-unknown-linux-gnu
  powerpc64le-unknown-linux-musl
  riscv32-wrs-vxworks
  riscv32e-unknown-none-elf
  riscv32em-unknown-none-elf
  riscv32emc-unknown-none-elf
  riscv32gc-unknown-linux-gnu
  riscv32gc-unknown-linux-musl
  riscv32i-unknown-none-elf
  riscv32im-risc0-zkvm-elf
  riscv32im-unknown-none-elf
  riscv32ima-unknown-none-elf
  riscv32imac-esp-espidf
  riscv32imac-unknown-none-elf
  riscv32imac-unknown-nuttx-elf
  riscv32imac-unknown-xous-elf
  riscv32imafc-esp-espidf
  riscv32imafc-unknown-none-elf
  riscv32imafc-unknown-nuttx-elf
  riscv32imc-esp-espidf
  riscv32imc-unknown-none-elf
  riscv32imc-unknown-nuttx-elf
  riscv64-linux-android
  riscv64-wrs-vxworks
  riscv64gc-unknown-freebsd
  riscv64gc-unknown-fuchsia
  riscv64gc-unknown-hermit
  riscv64gc-unknown-linux-gnu
  riscv64gc-unknown-linux-musl
  #riscv64gc-unknown-netbsd
  riscv64gc-unknown-none-elf
  riscv64gc-unknown-nuttx-elf
  riscv64gc-unknown-openbsd
  riscv64imac-unknown-none-elf
  riscv64imac-unknown-nuttx-elf
  s390x-unknown-linux-gnu
  s390x-unknown-linux-musl
  sparc-unknown-linux-gnu
  sparc-unknown-none-elf
  sparc64-unknown-linux-gnu
  sparc64-unknown-netbsd
  sparc64-unknown-openbsd
  sparcv9-sun-solaris
  thumbv4t-none-eabi
  thumbv5te-none-eabi
  thumbv6m-none-eabi
  thumbv6m-nuttx-eabi
  thumbv7a-pc-windows-msvc
  thumbv7a-uwp-windows-msvc
  thumbv7em-none-eabi
  thumbv7em-none-eabihf
  thumbv7em-nuttx-eabi
  thumbv7em-nuttx-eabihf
  thumbv7m-none-eabi
  thumbv7m-nuttx-eabi
  thumbv7neon-linux-androideabi
  thumbv7neon-unknown-linux-gnueabihf
  thumbv7neon-unknown-linux-musleabihf
  thumbv8m.base-none-eabi
  thumbv8m.base-nuttx-eabi
  thumbv8m.main-none-eabi
  thumbv8m.main-none-eabihf
  thumbv8m.main-nuttx-eabi
  thumbv8m.main-nuttx-eabihf
  wasm32-unknown-emscripten
  wasm32-unknown-unknown
  wasm32-wasip1
  wasm32-wasip1-threads
  wasm32-wasip2
  wasm32v1-none
  wasm64-unknown-unknown
  x86_64-apple-darwin
  x86_64-apple-ios
  x86_64-apple-ios-macabi
  x86_64-apple-tvos
  x86_64-apple-watchos-sim
  x86_64-fortanix-unknown-sgx
  x86_64-linux-android
  x86_64-pc-nto-qnx710
  x86_64-pc-solaris
  x86_64-pc-windows-gnu
  x86_64-pc-windows-gnullvm
  x86_64-pc-windows-msvc
  x86_64-unikraft-linux-musl
  x86_64-unknown-dragonfly
  x86_64-unknown-freebsd
  x86_64-unknown-fuchsia
  x86_64-unknown-haiku
  x86_64-unknown-hermit
  x86_64-unknown-hurd-gnu
  x86_64-unknown-illumos
  x86_64-unknown-l4re-uclibc
  x86_64-unknown-linux-gnu
  x86_64-unknown-linux-gnux32
  x86_64-unknown-linux-musl
  #x86_64-unknown-linux-none
  x86_64-unknown-linux-ohos
  x86_64-unknown-netbsd
  x86_64-unknown-none
  x86_64-unknown-openbsd
  x86_64-unknown-redox
  x86_64-unknown-trusty
  x86_64-unknown-uefi
  x86_64-uwp-windows-gnu
  x86_64-uwp-windows-msvc
  x86_64-win7-windows-msvc
  x86_64-wrs-vxworks
  x86_64h-apple-darwin
  #xtensa-esp32-espidf
  #xtensa-esp32-none-elf
  #xtensa-esp32s2-espidf
  #xtensa-esp32s2-none-elf
  #xtensa-esp32s3-espidf
  #xtensa-esp32s3-none-elf
)

# Migrate to 2018

for target in "${targets[@]}"; do
  echo "Migrating 2015 -> 2018 target $target"
  cargo fix --target "$target" -Zbuild-std=core --features extra_traits --allow-dirty --edition --broken-code 2> /dev/null
done

# Handle freebsd versions
freebsd_versions=(freebsd13 freebsd14 freebsd15)
for ver in "${freebsd_versions[@]}"; do
  echo "Migrating 2015 -> 2018 $ver"
  RUSTFLAGS="--cfg $ver" cargo fix --target i686-unknown-freebsd -Zbuild-std=core --features extra_traits --allow-dirty --edition --broken-code 2> /dev/null
  RUSTFLAGS="--cfg $ver" cargo fix --target x86_64-unknown-freebsd -Zbuild-std=core --features extra_traits --allow-dirty --edition --broken-code 2> /dev/null
done

perl -pi -077 -e 's/edition = "2015"/edition = "2018"/' Cargo.toml

# Migrate to 2021

for target in "${targets[@]}"; do
  echo "Migrating 2018 -> 2021 target $target"
  cargo fix --target "$target" -Zbuild-std=core --features extra_traits --allow-dirty --edition --broken-code 2> /dev/null
done

perl -pi -077 -e 's/edition = "2018"/edition = "2021"/' Cargo.toml

# Check
for target in "${targets[@]}"; do
  echo "Checking target $target"
  cargo check --target "$target" -Zbuild-std=core --features extra_traits || {
    echo "Failed for target $target"
    exit 1
  }
done

for ver in "${freebsd_versions[@]}"; do
  echo "Checking $ver"
  RUSTFLAGS="--cfg $ver" cargo check --target i686-unknown-freebsd -Zbuild-std=core --features extra_traits || {
    echo "Failed for i686 $freebsd_versions"
    exit 1
  }
  RUSTFLAGS="--cfg $ver" cargo check --target x86_64-unknown-freebsd -Zbuild-std=core --features extra_traits || {
    echo "Failed for x86_64 $freebsd_versions"
    exit 1
  }
done

# Fix formatting
ci/style.sh
```

</details>

tgross35 commented on libc-test/build.rs:
```diff
@@ -145,11 +145,41 @@ fn main() {
     // Avoid unnecessary re-building.
     println!("cargo:rerun-if-changed=build.rs");
 
+    if std::fs::exists("../target/src-hotfix").unwrap() {
```
Sorry, I just noticed one more thing: I think these need to use `CARGO_TARGET_DIR` since if that is overridden then we shouldn't write to `../target`.

eduardosm replied:
I used `OUT_DIR` instead, since I think it is more appropriate and Cargo won't set `CARGO_TARGET_DIR` if the user does not define it.

tgross35 reviewed (approved):
Awesome, `git ls-tree --full-tree -r HEAD . | git hash-object --stdin` reports 9bdf35a45c01d45498aa0bcaf8baaee3ce1d6d8a both when I run locally and when I check the latest commit here (bb8bbf23e206ecce56acbe42cc7f07fac5f18687). For some reason I had to pass `--lib` to `cargo fix` and `cargo check`, otherwise it wanted to build `std`.

Anyway I think this should be good, let's get it merged.

tgross35 commented:
Thanks for all the work here! I'll backport this and add a `.git-blame-ignore-revs` entry for the third commit.

tgross35 commented:
I think the older freebsd versions that we don't test in CI might have gotten skipped. I'll make sure everything gets synced up before this makes it to a release (fyi @asomers).

tgross35 commented:
Backported in https://github.com/rust-lang/libc/pull/4155#event-15445820840
